File: convex/functions/transaction.ts
import { mutation } from "../_generated/server";
import { v } from "convex/values";
import { validateTableAccess, validateWhereClause, sanitizeData } from "../lib/validators";
import { validateData } from "../lib/validation_rules";

/**
 * Transaction System for CFX-DB
 *
 * Provides atomic multi-operation batches to prevent money/item duplication exploits.
 * All operations in a transaction succeed or fail together.
 *
 * IMPORTANT: Convex mutations are already atomic within a single function.
 * This implementation leverages that guarantee by batching operations into one mutation.
 */

// Maximum operations per transaction (prevent abuse)
const MAX_OPERATIONS_PER_TRANSACTION = 10;

// Transaction timeout (5 seconds)
const TRANSACTION_TIMEOUT_MS = 5000;

/**
 * Operation types supported in transactions
 */
type TransactionOperation =
    | {
          type: "insert";
          table: string;
          data: any;
      }
    | {
          type: "update";
          table: string;
          where: any;
          data: any;
      }
    | {
          type: "delete";
          table: string;
          where: any;
      }
    | {
          type: "upsert";
          table: string;
          where: any;
          insert: any;
          update: any;
      };

/**
 * Execute atomic transaction
 *
 * All operations succeed or all fail - no partial commits.
 * This prevents money/item duplication exploits.
 */
export const executeTransaction = mutation({
    args: {
        operations: v.array(v.any()),
        timeout: v.optional(v.number()),
    },
    handler: async (ctx, args) => {
        const startTime = Date.now();

        // Validate operation count
        if (args.operations.length === 0) {
            throw new Error("Transaction must have at least one operation");
        }

        if (args.operations.length > MAX_OPERATIONS_PER_TRANSACTION) {
            throw new Error(
                `Transaction exceeds maximum ${MAX_OPERATIONS_PER_TRANSACTION} operations`
            );
        }

        // Validate all operations BEFORE executing ANY
        // This ensures we don't partially apply operations before hitting an error
        for (const op of args.operations) {
            if (!op.type || !op.table) {
                throw new Error("Each operation must have 'type' and 'table' fields");
            }

            // Validate table access
            validateTableAccess(op.table);

            // Validate where clauses
            if (op.where) {
                validateWhereClause(op.where);
            }

            // Validate data
            if (op.data) {
                const sanitized = sanitizeData(op.data);
                validateData(op.table, sanitized, op.type === "insert" ? "insert" : "update");
            }

            if (op.insert) {
                const sanitized = sanitizeData(op.insert);
                validateData(op.table, sanitized, "insert");
            }

            if (op.update) {
                const sanitized = sanitizeData(op.update);
                validateData(op.table, sanitized, "update");
            }
        }

        // Execute all operations atomically
        const results: any[] = [];

        for (const op of args.operations) {
            // Check timeout
            if (Date.now() - startTime > (args.timeout || TRANSACTION_TIMEOUT_MS)) {
                throw new Error("Transaction timeout exceeded");
            }

            let result: any;

            switch (op.type) {
                case "insert": {
                    const sanitized = sanitizeData(op.data);
                    const id = await ctx.db.insert(op.table as any, sanitized);
                    result = { type: "insert", id };
                    break;
                }

                case "update": {
                    if (!op.where) {
                        throw new Error("UPDATE requires where clause");
                    }

                    // Find matching records
                    let query = ctx.db.query(op.table);
                    if (op.where._id) {
                        // Optimize for _id lookups
                        const record = await ctx.db.get(op.where._id);
                        if (!record) {
                            result = { type: "update", updated: 0 };
                            break;
                        }

                        const sanitized = sanitizeData(op.data);
                        await ctx.db.patch(record._id, sanitized);
                        result = { type: "update", updated: 1 };
                    } else {
                        // General where clause (less efficient, but necessary)
                        const records = await query.collect();
                        const matching = records.filter((r: any) => {
                            // Simple equality check for now
                            for (const [key, value] of Object.entries(op.where)) {
                                if (r[key] !== value) return false;
                            }
                            return true;
                        });

                        if (matching.length === 0) {
                            result = { type: "update", updated: 0 };
                            break;
                        }

                        const sanitized = sanitizeData(op.data);
                        for (const record of matching) {
                            await ctx.db.patch(record._id, sanitized);
                        }

                        result = { type: "update", updated: matching.length };
                    }
                    break;
                }

                case "delete": {
                    if (!op.where) {
                        throw new Error("DELETE requires where clause");
                    }

                    // Find and delete matching records
                    let query = ctx.db.query(op.table);
                    if (op.where._id) {
                        // Optimize for _id lookups
                        const record = await ctx.db.get(op.where._id);
                        if (!record) {
                            result = { type: "delete", deleted: 0 };
                            break;
                        }

                        await ctx.db.delete(record._id);
                        result = { type: "delete", deleted: 1 };
                    } else {
                        const records = await query.collect();
                        const matching = records.filter((r: any) => {
                            for (const [key, value] of Object.entries(op.where)) {
                                if (r[key] !== value) return false;
                            }
                            return true;
                        });

                        for (const record of matching) {
                            await ctx.db.delete(record._id);
                        }

                        result = { type: "delete", deleted: matching.length };
                    }
                    break;
                }

                case "upsert": {
                    if (!op.where) {
                        throw new Error("UPSERT requires where clause");
                    }

                    // Try to find existing record
                    let query = ctx.db.query(op.table);
                    const records = await query.collect();
                    const existing = records.find((r: any) => {
                        for (const [key, value] of Object.entries(op.where)) {
                            if (r[key] !== value) return false;
                        }
                        return true;
                    });

                    if (existing) {
                        // Update
                        const sanitized = sanitizeData(op.update);
                        await ctx.db.patch(existing._id, sanitized);
                        result = { type: "upsert", action: "updated", id: existing._id };
                    } else {
                        // Insert
                        const sanitized = sanitizeData(op.insert);
                        const id = await ctx.db.insert(op.table as any, sanitized);
                        result = { type: "upsert", action: "inserted", id };
                    }
                    break;
                }

                default:
                    throw new Error(`Unknown operation type: ${op.type}`);
            }

            results.push(result);
        }

        return {
            success: true,
            operations: results.length,
            results,
            duration: Date.now() - startTime,
        };
    },
});

/**
 * IMPORTANT: This is a GENERIC transaction system.
 * Application-specific logic (like money transfers) should be implemented
 * in YOUR gamemode using the executeTransaction function above.
 *
 * Example usage in your gamemode:
 *
 * local tx = DB.transaction()
 *     :update('players', {where = {identifier = from}, set = {money = money - 100}})
 *     :update('players', {where = {identifier = to}, set = {money = money + 100}})
 *     :insert('transactions', {from = from, to = to, amount = 100})
 *     :commit(callback)
 *
 * This keeps cfx-db generic and reusable across different gamemodes.
 */

File: lua/transaction.lua
--[[
    Transaction Builder for CFX-DB

    Provides a fluent API for building atomic transactions.
    All operations in a transaction succeed or fail together.

    NOTE: Methods are attached directly to each instance (not via metatable)
    because FiveM strips metatables when tables cross resource boundaries.

    Example usage:
        local tx = exports['cfx-db']:transaction()
            :update('players', {where = {identifier = license}, set = {money = money - 100}})
            :update('players', {where = {identifier = target}, set = {money = money + 100}})
            :insert('transactions', {from = license, to = target, amount = 100, time = os.time()})
            :commit(function(success, result)
                if success then
                    print("Transaction committed!")
                else
                    print("Transaction failed:", result)
                end
            end)
]]

-- Detect which cfx-db resource is running (dev: cfx-db-dev, prod: cfx-db)
local cfxDbResource = GetResourceState('cfx-db-dev') == 'started' and 'cfx-db-dev' or 'cfx-db'
local log = Log.scoped("transaction")

---Create a new transaction instance with methods directly attached
---@return table Transaction instance
local function create()
    local tx = {
        operations = {}
    }

    ---Add an INSERT operation
    ---@param tableName string Table name
    ---@param data table Data to insert
    ---@return table self for chaining
    function tx:insert(tableName, data)
        if type(tableName) ~= "string" then
            error("Transaction:insert() - tableName must be a string")
        end
        if type(data) ~= "table" then
            error("Transaction:insert() - data must be a table")
        end

        table.insert(self.operations, {
            type = "insert",
            table = tableName,
            data = data
        })
        return self
    end

    ---Add an UPDATE operation
    ---@param tableName string Table name
    ---@param params table {where = {...}, set = {...}}
    ---@return table self for chaining
    function tx:update(tableName, params)
        if type(tableName) ~= "string" then
            error("Transaction:update() - tableName must be a string")
        end
        if type(params) ~= "table" then
            error("Transaction:update() - params must be a table")
        end
        if not params.where then
            error("Transaction:update() - params.where is required")
        end
        if not params.set then
            error("Transaction:update() - params.set is required")
        end

        table.insert(self.operations, {
            type = "update",
            table = tableName,
            where = params.where,
            data = params.set
        })
        return self
    end

    ---Add a DELETE operation
    ---@param tableName string Table name
    ---@param where table Where clause
    ---@return table self for chaining
    function tx:delete(tableName, where)
        if type(tableName) ~= "string" then
            error("Transaction:delete() - tableName must be a string")
        end
        if type(where) ~= "table" then
            error("Transaction:delete() - where must be a table")
        end

        table.insert(self.operations, {
            type = "delete",
            table = tableName,
            where = where
        })
        return self
    end

    ---Add an UPSERT operation
    ---@param tableName string Table name
    ---@param params table {where = {...}, insert = {...}, update = {...}}
    ---@return table self for chaining
    function tx:upsert(tableName, params)
        if type(tableName) ~= "string" then
            error("Transaction:upsert() - tableName must be a string")
        end
        if type(params) ~= "table" then
            error("Transaction:upsert() - params must be a table")
        end
        if not params.where then
            error("Transaction:upsert() - params.where is required")
        end
        if not params.insert then
            error("Transaction:upsert() - params.insert is required")
        end
        if not params.update then
            error("Transaction:upsert() - params.update is required")
        end

        table.insert(self.operations, {
            type = "upsert",
            table = tableName,
            where = params.where,
            insert = params.insert,
            update = params.update
        })
        return self
    end

    ---Commit the transaction
    ---@param callback function Callback function(success, result)
    function tx:commit(callback)
        if type(callback) ~= "function" then
            error("Transaction:commit() - callback must be a function")
        end

        if #self.operations == 0 then
            callback(false, "Transaction has no operations")
            return
        end

        if #self.operations > 10 then
            callback(false, "Transaction exceeds maximum 10 operations")
            return
        end

        exports[cfxDbResource]:executeTransaction(self.operations, callback)
    end

    ---Get the number of operations in the transaction
    ---@return number
    function tx:size()
        return #self.operations
    end

    ---Clear all operations
    ---@return table self for chaining
    function tx:clear()
        self.operations = {}
        return self
    end

    return tx
end

-- Expose transaction API (GENERIC - no application logic)
_G.DBTransaction = {
    create = create,
}

-- Add to main DB namespace
if DB then
    DB.transaction = create
    log.debug("Transaction API registered")
end

return create
